src/Symfony/Component/Mime/Tests/Part/SMimePartTest.php

```diff
@@ -0,0 +1,91 @@
+use Symfony\Component\Mime\Header\ParameterizedHeader;
+use Symfony\Component\Mime\Part\SMimePart;
+
+class SMimePartTest extends TestCase
+{
+    public function testConstructor()
+    {
+        $p = new SMimePart('content', 'multipart', 'signed', []);
+        $this->assertEquals('content', $p->bodyToString());
+        $this->assertEquals('content', implode('', iterator_to_array($p->bodyToIterable())));
+        // bodyToIterable() can be called several times
+        $this->assertEquals('content', implode('', iterator_to_array($p->bodyToIterable())));
+        $this->assertEquals('multipart', $p->getMediaType());
+        $this->assertEquals('signed', $p->getMediaSubType());
+    }
+
+    public function testConstructorWithIterable()
+    {
+        $iterable = $this->getIterable();
+
+        $p = new SMimePart($iterable, 'multipart', 'signed', []);
+        $this->assertEquals('content', $p->bodyToString());
+        $this->assertEquals('content', implode('', iterator_to_array($p->bodyToIterable())));
+        // bodyToIterable() can be called several times
+        $this->assertEquals('content', implode('', iterator_to_array($p->bodyToIterable())));
+        $this->assertEquals('multipart', $p->getMediaType());
+        $this->assertEquals('signed', $p->getMediaSubType());
+    }
+
+    public function testConstructorWithInvalidIterable()
+    {
+        $iterable = $this->getIterable();
+        // We using this method for close a generator which should throws: Exception : Cannot traverse an already closed generator
+        iterator_to_array($iterable);
+
+        $p = new SMimePart($iterable, 'multipart', 'signed', []);
+        $this->assertEquals('', $p->bodyToString());
+        $this->assertEquals('', implode('', iterator_to_array($p->bodyToIterable())));
+        // bodyToIterable() can be called several times
+        $this->assertEquals('', implode('', iterator_to_array($p->bodyToIterable())));
+        $this->assertEquals('multipart', $p->getMediaType());
+        $this->assertEquals('signed', $p->getMediaSubType());
+    }
+
+    public function testConstructorWithNonStringOrIterable()
+    {
+        $this->expectException(\TypeError::class);
+        new SMimePart(new \stdClass(), 'multipart', 'signed', []);
+    }
+
+    public function testHeaders()
+    {
+        $p = new SMimePart('content', 'multipart', 'signed', []);
+        $this->assertEquals(new Headers(
+            new ParameterizedHeader('Content-Type', 'multipart/signed', [])
+        ), $p->getPreparedHeaders());
+
+        $p = new SMimePart('content', 'multipart', 'signed', ['charset' => 'utf-8']);
+        $this->assertEquals(new Headers(
+            new ParameterizedHeader('Content-Type', 'multipart/signed', ['charset' => 'utf-8'])
+        ), $p->getPreparedHeaders());
+    }
+
+    private function getIterable(): iterable
+    {
+        $f = fopen('php://memory', 'r+', false);
+        fwrite($f, 'content');
+        rewind($f);
```

Comment:
Do we really need to use a file here? Can't we just:

```
private function getIterable(): iterable
{
    yield 'con';
    yield 'tent';
}
```

Reply:
Yes, we can :)